src/deroff.rs

```diff
@@ -538,12 +537,311 @@ impl Deroffer {
         unimplemented!()
     }
 
+    fn quoted_arg(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn request_or_macro(&mut self) -> bool {
+        // self.skip_char();
+        self.s = self.skip_char(&self.s, None).to_owned();
+
+        let s0 = &self.s[1..2];
+
+        match s0 {
+            "\\" => {
+                // if self.str_at(1) == Some('"') {
+                if Self::str_at(&self.s, 1) == "\"" {
+                    // self.condputs("\n");
+                    return true;
+                }
+            }
+            "[" => {
+                self.refer = true;
+                // self.condputs("\n");
+                return true;
+            }
+            "]" => {
+                self.refer = false;
+                // self.skip_char();
+                self.s = self.skip_char(&self.s, None).to_owned();
```

Comment:
Why do we have `to_owned()` for `String` everywhere? Isn't this way slower than parsing just `&str`?

Reply:
`self.s` is a `String`. #69 exists, I don't like the interfaces for those functions because they make you convert into a `&str` and then back to `String`. Modifying self.s in place rather than converting it to a `&str` and then setting it to the result converted to a `String` would be significantly faster and cleaner

Reply:
I was looking at this a little bit. I think there are tons of opportunities to get rid of some of the worst of the cloning. One thing to note for some of the cases that show up in this PR though, is that the signature of `condputs` from the perspective of this PR should likely be `condputs(&mut self, _: &str)` instead of `condputs(&self, _: &str)`, which will then prevent you from passing in an `&str` that is borrowed from `self`.

@pickfire, we might initially have something very slow by Rust standards. I'm _doubtful_ it will be any slower than the original—I'd expect it to be at least a little faster. However, we still don't have something fully working in the `deroff` module. I'd like to have that plus a comprehensive set of unit tests for `deroff`'s behavior before we revisit to optimize.

Reply:
I should have known that, I feel like this will lead to more clones than before. I don't know how to handle stuff like in `word` without a clone

```rust
while let Some(m) = self.g_re_word.find(&self.s) {
    self.condputs(m.as_str());
```

Reply:
Yeah, if any I believe we can still discuss on the design of parsing later on. A rough idea is to just keep the original `String` or `&str` and use an index to iterate through it.